sycl-rfc-unnamed-lambda.md

```diff
@@ -0,0 +1,115 @@
+[Appendix B.2, "Reduced feature set"](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:feature-sets.reduced).
+Per
+[section 5.6, "Preprocessor directives and macros"](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_preprocessor_directives_and_macros),
+implementations that do not support unnamed SYCL kernel functions are required to predefine
+the `SYCL_FEATURE_SET_REDUCED` macro with the value `1` and to not define `SYCL_FEATURE_SET_FULL`.
+SYCL programs that use unnamed SYCL kernel functions are therefore more limited in their
+portability.
+
+Support for unnamed SYCL kernel functions is more easily achieved when the host compiler
+is SYCL-aware since this allows the host and device compilers to use more surrounding context
+to generate a stable kernel name than is incorporated in the existing C++ name mangling schemes
+used for lambdas.
+Clang already supports a `__builtin_sycl_unique_stable_name` builtin function to assist with
+production of a stable kernel name.
+Modulo defects in its design and implementation, this suffices to provide full support of
+unnamed SYCL kernel functions when Clang is used as both the host and device compiler.
+
+Support for SYCL-unaware host compilers could be provided by preprocessing a SYCL translation
+unit, instrumenting the resulting preprocessed output with a SYCL-aware tool to insert stable
+kernel names in calls to SYCL kernel invocation functions, and then passing the result to
+the SYCL-unaware host compiler.
+This approach has several downsides including:
+- Loss of preprocessing information during host compilation. This may affect when diagnostics
+  are issued and their presentation form.
+- Loss of precise source location information. This may affect diagnostic presentation form
+  and the quality of debugging information.
+- Performing the instrumentation would be technically challenging as it would require the
+  ability to consume preprocessed output intended for the host compiler while also being able
+  to parse the code sufficiently well to deterministically generate kernel names that match
+  those produced by the device compiler or to correlate kernel invocations with a table of
+  kernel names produced by the device compiler.
+
+At present, the `__builtin_sycl_unique_stable_name` builtin function does not generate
+matching names during host and device compilation for the following example.
+This is presumed to be either a defect in its implementation or a limitation of its design.
+Running this program results in a run-time error due to a failure to resolve kernel names.
+It is not entirely clear whether the SYCL specification requires a conforming implementation
+to support this example.
+An issue has been filed against the SYCL specification to request clarification.
+See [issue #434, "Correspondence of unnamed lambdas as kernels across host and device compilation"](https://github.com/KhronosGroup/SYCL-Docs/issues/454).
```

Comment:
Can we replace these sentences with something like this instead?

> This is a limitation we would like to avoid, and we plan to investigate possible solutions.

I think this is aligned with your list of proposed bullet points below.

I think the comment about the unclarity of the spec isn't really that accurate.  In its current wording, the spec does not prohibit use of `__SYCL_DEVICE_ONLY__` along the lines you show, so the assumption must be that this code is required to work.  I think the burden is therefore on implementors to either make this construct work, or to make a case to the SYCL WG that it is unreasonable to support this code.

Reply:
Yes, good suggestion. Since the intent to address cases like this is already present in the proposal list, I just went ahead and removed those lines. I reworded the paragraph a bit as well.